File: Coder/CSharp/CSharpParser.cs
using System;
using System.IO;
using Antlr.Runtime;
using Antlr.Runtime.Tree;

namespace Coder.CSharp
{
    public class CSharpParser
    {
        public void ParseDirectory(string path, string searchPattern, IQueryResults results)
        {
            if (path == string.Empty)
                path = ".";

            ProcessDirectory(path, searchPattern, results);

            ProcessFiles(path, searchPattern, results);
        }

        private void ProcessFiles(string path, string searchPattern, IQueryResults results)
        {
            foreach (var f in Directory.GetFiles(path, searchPattern))
            {
                var wait = searchPattern.IndexOfAny(new[] {'*', '?'}) != -1;
                if (Args.IsFlagSet("-q")) // Quick set?
                    wait = false;

                ParseFile(f, wait, results);
            }
        }

        private void ProcessDirectory(string path, string searchPattern, IQueryResults results)
        {
            foreach (var subdir in Directory.GetDirectories(path))
            {
                ParseDirectory(subdir, searchPattern, results);
            }
        }

        public void ParseFile(string fileName, bool wait, IQueryResults results)
        {
            results.ProcessingFile(fileName);

            var tokens = CreateLexer<PreProcessor>(fileName);
            var p = new csParser(tokens) {TreeAdaptor = new CommonTreeAdaptor()};
            CommonTree tree = null;

            var parserResult = p.compilation_unit();

            if (parserResult != null) tree = (CommonTree) parserResult.Tree;

            if (tree == null)
            {
                if (tokens.Count > 0)
                {
                    results.ErrorProcessingFile(fileName);
                }
            }
            else
            {
                var nodes = new CommonTreeNodeStream(tree) {TokenStream = tokens};
                FindEmptyCatchBlocks(fileName, nodes, results);
            }
        }

        private void FindEmptyCatchBlocks(string fileName, CommonTreeNodeStream nodes, IQueryResults results)
        {
            try
            {
                var current = nodes.NextElement() as CommonTree;
                while (!nodes.IsEndOfFile(current))
                {
                    if (TokenMatches(current, "catch"))
                    {
                        var catchToken = current.Token;
                        results.CatchFound(catchToken);

                        current = nodes.NextElement() as CommonTree;

                        if (TokenMatches(current, "{"))
                        {
                            var openingBrace = current.Token;
                            current = nodes.NextElement() as CommonTree;
                            if (TokenMatches(current, "}"))
                            {
                                results.EmptyCatchFound(catchToken, openingBrace, current.Token);
                            }
                        }
                    }
                    current = nodes.NextElement() as CommonTree;
                }
            }
            catch (Exception)
            {
                results.ErrorProcessingFile(fileName);
            }
        }

        private static bool TokenMatches(CommonTree ct, string text)
        {
            return ct != null && ct.Token != null && ct.Token.Text == text;
        }

        public CommonTokenStream CreateLexer<TL>(string fileName)
            where TL : Lexer, new()
        {
            var inputFileName = fileName;

            if (!Path.IsPathRooted(inputFileName))
                inputFileName = Path.Combine(Environment.CurrentDirectory, inputFileName);

            ICharStream input = new ANTLRFileStream(inputFileName);
            var lex = new TL {CharStream = input};

            return new CommonTokenStream(lex);
        }
    }
}

File: CodeR.CSharp.Parser/src/scala/CodeR/CSharp/Parser/CSharpParser.scala
package CodeR.CSharp.Parser

import CodeR.CSharp.csParser
import org.antlr.runtime.tree._
import org.antlr.runtime.{ANTLRFileStream, CommonTokenStream}
import CodeR.QueryResults

class CSharpParser {
  def ParseFile(fileName: String, wait: Boolean, results: QueryResults) {
    results.processingFile(fileName)

    val tokens = createLexer(fileName)
    val p = new csParser(tokens)
    p.setTreeAdaptor(new CommonTreeAdaptor())

    val parserResult = p.compilation_unit()

    if (parserResult != null) {
      val tree = parserResult.getTree.asInstanceOf[CommonTree]

      if (tree == null) {
        if (tokens.size > 0) {
          results.errorProcessingFile(fileName)
        }
      }
      else {
        val nodes = new CommonTreeNodeStream(tree)
        nodes.setTokenStream(tokens)
        FindEmptyCatchBlocks(fileName, nodes, results)
      }
    }
  }

  def FindEmptyCatchBlocks(fileName: String, nodes: CommonTreeNodeStream, results: QueryResults) {
    try {
      var current = nodes.nextElement.asInstanceOf[CommonTree]
      while (!nodes.isEOF(current)) {
        if (TokenMatches(current, "catch")) {
          val catchToken = current.getToken
          results.catchFound(catchToken)

          current = nodes.nextElement.asInstanceOf[CommonTree]

          if (TokenMatches(current, "{")) {
            val openingBrace = current.getToken
            current = nodes.nextElement().asInstanceOf[CommonTree]
            if (TokenMatches(current, "}")) {
              results.emptyCatchFound(catchToken, openingBrace, current.getToken)
            }
          }
        }
        current = nodes.nextElement.asInstanceOf[CommonTree]
      }
    }
    catch {
      case e: Exception => results.errorProcessingFile(fileName)
    }
  }

  def TokenMatches(ct: CommonTree, text: String): Boolean = ct != null && ct.getToken != null && ct.getToken.getText == text


  def createLexer(fileName: String): CommonTokenStream = {
    val inputFileName = fileName

    val input = new ANTLRFileStream(inputFileName)

    return new CommonTokenStream(new PreProcessor(input))
  }
}
